prepreduce/prepreduce.py

```diff
@@ -0,0 +1,395 @@
+#!/usr/bin/env python3
+# developed + tested on python 3.11.2
+
+import argparse
+import os
+from pathlib import Path
+import re
+import subprocess
+import textwrap
+
+from icecream import ic  # TODO: make zero-dependency
+
+DEBUG = False
+ic.disable()
+
+VERSION = "0.1"
+
+
+def initial_make(
+    build_command: list[str], target: Path, path_to_linux: Path, force_rm: bool
+) -> str:
+    """
+    Step 1: make the target.o file and scrape the original compiler invocation
+    (just clang for now)
+    """
+
+    absolute_target: Path = path_to_linux / target
+    if absolute_target.exists():
+        double_check_removal_with_user(file=absolute_target, force_rm=force_rm)
+
+    build_command_with_target = build_command[::] + [str(target)]
+
+    info(f"Running initial make with `{build_command} {target}`")
+    ic(build_command_with_target)
+
+    make_result = subprocess.run(
+        build_command_with_target,
+        text=True,
+        cwd=path_to_linux,
+        capture_output=True,
+    )
+
+    if len(make_result.stdout) and DEBUG:
+        with open("init_make.out", "w") as fd:
+            fd.write(make_result.stdout)
+        ic(f"make stdout written to init_make.out")
+
+    if make_result.returncode != 0:
+        error(f"Initial make failed: {make_result.stderr}")
+    else:
+        success("Initial make success")
+
+    return make_result.stdout
+
+
+def make_dot_i_file(
+    cc_invocation: list[str], path_to_linux: Path, output_dir: Path
+) -> Path:
+    """
+    Make the preprocessed .i file which can be compiled sans build system
+    """
+    cc = cc_invocation[::]
+
+    try:
+        dash_c_idx = cc.index("-c")
+        dash_o_idx = cc.index("-o")
+    except ValueError:
+        error(
+            "Somehow there is no -c and/or -o flag in the main invocation "
+            "(not your fault, it's prepreduce's fault) quitting."
+        )
+        exit(1)  # redundant exit, but it makes Pyright be quiet about unbound vars
+
+    output_target_idx = dash_o_idx + 1
+
+    cc[dash_c_idx] = "-E"
+    stem = Path(cc[output_target_idx]).stem
+    local_path_for_dot_i_file = Path(os.getcwd()) / Path(stem).with_suffix(".i")
+    ic(local_path_for_dot_i_file)
+
+    if (loc := local_path_for_dot_i_file).exists():
+        new_name = loc.with_name(".old." + loc.stem + loc.suffix)
+        warn(f"{loc} already exists. Renaming it to {new_name}")
+        os.rename(local_path_for_dot_i_file, output_dir / new_name)
```

Comment:
This can be shorter with `.rename()` from `pathlib`:

```python
local_path_for_dot_i_file.rename(output_dir / new_name)
```

Additionally, if I am right about the `output_dir` comment above, I do not think that you need `output_dir / `, as `new_name` will be in `output_dir` already.